app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/CalendarController.kt

```diff
@@ -0,0 +1,267 @@
+     */
+    fun hasLectures() = calendarDao.hasLectures()
+
+    /**
+     * Add a lecture to the blacklist of a widget
+     *
+     * @param widgetId the Id of the widget
+     * @param lecture  the title of the lecture
+     */
+    fun addLectureToBlacklist(widgetId: Int, lecture: String) {
+        widgetsTimetableBlacklistDao.insert(WidgetsTimetableBlacklist(widgetId, lecture))
+    }
+
+    /**
+     * Remove a lecture from the blacklist of a widget
+     *
+     * @param widgetId the Id of the widget
+     * @param lecture  the title of the lecture
+     */
+    fun deleteLectureFromBlacklist(widgetId: Int, lecture: String) {
+        widgetsTimetableBlacklistDao.delete(WidgetsTimetableBlacklist(widgetId, lecture))
+    }
+
+    /**
+     * Get all lectures and the information whether they are on the blacklist for the given widget
+     *
+     * @param widgetId the Id of the widget
+     * @return A cursor containing a list of lectures and the is_on_blacklist flag
+     */
+    fun getLecturesForWidget(widgetId: Int): List<CalendarItem> {
+        val lectures = calendarDao.getLecturesInBlacklist(widgetId.toString()).toMutableList()
+        lectures.forEach {
+            it.blacklisted = true
+        }
+        lectures.addAll(calendarDao.getLecturesNotInBlacklist(widgetId.toString()))
+        return lectures
+    }
+
+    /**
+     * Gets the event by its id and duplicates of this event with different locations.
+     * The first item is the one with the given id.
+     */
+    fun getCalendarItemAndDuplicatesById(id: String) = calendarDao.getCalendarItemsById(id)
+
+    fun scheduleNotifications(events: List<Event>) {
+        // Be responsible when scheduling alarms. We don't want to exceed system resources
+        // By only using up half of the remaining resources, we achieve fair distribution of the
+        // remaining usable notifications
+        val maxNotificationsToSchedule = NotificationScheduler.maxRemainingAlarms(context) / 2
+
+        val notifications = events.filter { it.isFutureEvent }
+                .mapNotNull { it.toNotification(context) }
+                .take(maxNotificationsToSchedule)
+
+        val scheduler = NotificationScheduler(context)
+        scheduler.schedule(notifications)
+    }
+
+    fun importCalendar(events: List<Event>) {
+        // Cleanup cache before importing
+        removeCache()
+
+        // Import the new events
+        try {
+            replaceIntoDb(events)
+        } catch (e: Exception) {
+            Utils.log(e)
+        }
+        SyncManager(context).replaceIntoDb(Const.SYNC_CALENDAR_IMPORT)
+    }
+
+    /**
+     * Removes all cache items
+     */
+    private fun removeCache() {
+        calendarDao.flush()
+    }
+
+    private fun replaceIntoDb(events: List<Event>) {
+        val items = ArrayList<CalendarItem>()
+        for (event in events) {
+            if (event.id == null || event.id.isEmpty()) {
```

Comment:
These exceptions are not actually handled, but just logged. Maybe simply skip the events with invalid IDs and titles?